Add selective logging

matthiasr opened the pull request:
In high-volume services, it is not always useful to log all requests.
Allow specifying a classifying function that can inspect the event to
decide whether it should be logged.

For convenience, keep the existing behavior under the existing names, by
supplying a trivial classifier.

matthiasr committed:
go mod init

Signed-off-by: Matthias Rampke <matthias@rampke.de>

matthiasr committed:
Relax timing requirement for HTTP logging

Starting with Go 1.16, testing for millisecond-exact processing times
is quite prone to flakiness; I have seen this in other places as well.

Relax the timing requirement to be ±50% of the test duration – this is
exact enough to check whether the logging is reasonable.

Signed-off-by: Matthias Rampke <matthias@rampke.de>

matthiasr committed:
Add selective logger

In high-volume services, it is not always useful to log all requests.
Allow specifying a classifying function that can inspect the event to
decide whether it should be logged.

For convenience, keep the existing behavior under the existing names, by
supplying a trivial classifier.

Signed-off-by: Matthias Rampke <matthias@rampke.de>

matthiasr commented:
To make the tests pass with Go 1.16 reliably, I had to relax the timing requirements a fair bit; I hope that's acceptable.

telendt commented:
Is this `report.SelectiveJSONMiddleware` really needed though? This (and many other similar cases) could be solved with simple "conditional middleware":

```go
type ShouldForward func(*http.Request) bool

func ConditionalMiddleware(predicate ShouldForward, first, second http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if predicate(r) {
			first.ServeHTTP(w, r)
		} else {
			second.ServeHTTP(w, r)
		}
	})
}
```

The only thing to do is to change your "should forward" classifier so that it works on request instance. The first handler is your handler decorated with `report.JSON` and the second is the non-decorated one.

matthiasr commented:
This form doesn't work because we can only decide whether to log _after_ a request has been handled. The middleware needs to be in the call path so that it can do (or not do) something on the way back up the stack.

telendt commented:
> This form doesn't work because we can only decide whether to log *after* a request has been handled.

Ah, so you want to decide based on properties like ms, status or size? Now I get it.

I believe the problem comes from the fact `report.JSON` does too much - constructs the event instance from request and response, and then encodes that and writes the resulting bytes to the writer. If that event instance was passed to a "reporter" instead then you could do what you want there (filter) but it would also enable other use cases (like using different serializers or writing different events to different writers).

matthiasr committed:
Generalize report event handling

Introduce a Reporter interface that can be implemented in multiple ways.

Provide two Reporter implementations:

* one that logs JSON to a writer, as the monolithic JSONMiddleware did
* one that filters events based on a provided function, and calls
  another reporter selectively

Move the HTTP handling into a separate middleware that reports events to
any Reporter. Re-implement the JSONMiddleware using it.

Remove the selective JSON middleware again, it is now a one-liner that
can be constructed as needed. Do that in the tests to demonstrate how it
all fits together.

Signed-off-by: Matthias Rampke <matthias@rampke.de>

matthiasr commented:
@telendt like this?
